src/ast/irbuilderbpf.cpp

```diff
@@ -392,33 +388,33 @@ Value *IRBuilderBPF::CreateMapLookupElem(Value *ctx,
       "map_lookup_cond");
   CreateCondBr(condition, lookup_success_block, lookup_failure_block);
 
+  assert(value->getType()->isPointerTy());
+  llvm::Type *value_ptr_ty = value->getType()->getPointerElementType();
```

Comment:
The name `value_ptr_ty` is a little confusing for me - it's the element (i.e. pointee) type not the pointer type, right?

As for the functionality, bpftrace stores all integers in maps as 64-bit: https://github.com/iovisor/bpftrace/blob/master/src/ast/passes/semantic_analyser.cpp#L3232

Does this function not always return Int64Ty because of this?

On another note, apparently getPointerElementType is being deprecated: https://llvm.org/doxygen/classllvm_1_1Type.html#ae61f17d62c8a79bc5cffa63a2f152311
So if it's actually doing something other than always returning Int64Ty then we should try to find another way of doing it.

Reply:
`value_ptr_ty` is equivalent to `GetType(type)` here. I agree it's confusing; perhaps something like this would be better?

```
llvm::Type *value_ty = GetType(type);
AllocaInst *value = CreateAllocaBPF(value_ty, "lookup_elem_val");
// [...]
```

> Does this function not always return Int64Ty because of this?

Case where `type.IsIntTy() == true` should be unaffected; the issue here were maps containing pointers (for example, in `@a[0] = (struct Foo *) arg0` the map contains native pointers, so 4-byte values on arm32).

Hm, let me think some more whether it might be better to store pointers in maps as 64-bit ints. My concern originally was that this could lead to too many special cases (already had to do something similar in [ffebc5c](https://github.com/iovisor/bpftrace/pull/2361/commits/ffebc5c876892cba3d99bce36affdb494f258c59) to get around a BPF codegen issue).

> On another note, apparently getPointerElementType is being deprecated

I saw a comment in one of the LLVM headers about moving to opaque pointers (after which `getPointerElementType` will no longer work). We can avoid it in this method with the code above, but there are several other places in the codebase that would need to be updated. Not sure if there's a good alternative other than explicitly tracking T every time we create pointer-to-T, which might require quite a bit of refactoring. :/